### PR TITLE
[DDW-512] Implement dynamic TTL for HW transactions

### PR DESCRIPTION
**Pull request by tomothespian**
This PR Introduces a dynamic TTL for HW transactions.

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1612271536128400)
- [ ] "Send" transaction with Ledger and Trezor devices
- [ ] "Delegate" with Ledger and Trezor devices

---

## Test Cases

**TREZOR - Scenario 1 - Send ADA from Trezor T Wallet**

```Gherkin
Given a Trezor T wallet is added to Daedalus
And a Shelley wallet is added to Daedalus
And both wallets are synced
And the Trezor T wallet balance is >=2
When I access the "Send" screen for the Trezor T wallet
And I paste an address from the Shelley wallet on the "Receiver" input
And I insert a valid amount on the "Amount" input
And I click on "Next" and confirm the transaction
And I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the transaction is performed
And the transaction can be visible on both wallets
```

**TREZOR - Scenario 2 - Send ADA from Trezor T to Byron wallet**

```Gherkin
Given a Trezor T wallet is added to Daedalus
And a Byron wallet is added to Daedalus
And both wallets are synced
And the Trezor T wallet balance is >=2
When I access the "Send" screen for the Shelley wallet
And I paste an address from the Byron wallet on the "Receiver" input
And I insert a valid amount on the "Amount" input
And I click on Next"
And I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the transaction is performed
And the transaction can be visible on both wallets
```

**TREZOR - Scenario 3 - Redelegate Trezor T Wallet**

```Gherkin
Given a Trezor T wallet with at least 10 ADA is added
And the wallet is delegated
And both Daedalus and the wallet are synced
And the "Delegation Center" screen on "Staking Center" is displayed
When I click on "Redelegate" button for the wallet
And I follow the instructions on the wizard to redelegate the wallet
And I plug, unlock and export the public key on the device
And I confirm the delegation
Then the wallet is displayed as delegated to the new stake pool in the epoch number = current epoch + 2
```

**LEDGER NANO S - Scenario 1 - Send ADA from Ledger Nano S Wallet**

```Gherkin
Given a Ledger Nano S wallet is added to Daedalus
And a Shelley wallet is added to Daedalus
And both wallets are synced
And the Ledger Nano S wallet balance is >=2
When I access the "Send" screen for the Ledger Nano S wallet
And I paste an address from the Shelley wallet on the "Receiver" input
And I insert a valid amount on the "Amount" input
And I click on "Next" and confirm the transaction
And I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the transaction is performed
And the transaction can be visible on both wallets
```

**LEDGER NANO S - Scenario 2 - Send ADA from Ledger Nano S to Byron wallet**

```Gherkin
Given a Ledger Nano S wallet is added to Daedalus
And a Byron wallet is added to Daedalus
And both wallets are synced
And the Ledger Nano S wallet balance is >=2
When I access the "Send" screen for the Shelley wallet
And I paste an address from the Byron wallet on the "Receiver" input
And I insert a valid amount on the "Amount" input
And I click on Next"
And I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the transaction is performed
And the transaction can be visible on both wallets
```

**LEDGER NANO S - Scenario 3 - Redelegate Ledger Nano S Wallet**

```Gherkin
Given a Ledger Nano S wallet with at least 10 ADA is added
And the wallet is delegated
And both Daedalus and the wallet are synced
And the "Delegation Center" screen on "Staking Center" is displayed
When I click on "Redelegate" button for the wallet
And I follow the instructions on the wizard to redelegate the wallet
And I plug, unlock and export the public key on the device
And I confirm the delegation
Then the wallet is displayed as delegated to the new stake pool in the epoch number = current epoch + 2
```

**LEDGER NANO X - Scenario 1 - Send ADA from Ledger Nano X Wallet**

```Gherkin
Given a Ledger Nano X wallet is added to Daedalus
And a Shelley wallet is added to Daedalus
And both wallets are synced
And the Ledger Nano X wallet balance is >=2
When I access the "Send" screen for the Ledger Nano X wallet
And I paste an address from the Shelley wallet on the "Receiver" input
And I insert a valid amount on the "Amount" input
And I click on "Next" and confirm the transaction
And I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the transaction is performed
And the transaction can be visible on both wallets
```

**LEDGER NANO X - Scenario 2 - Send ADA from Ledger Nano X to Byron wallet**

```Gherkin
Given a Ledger Nano x wallet is added to Daedalus
And a Byron wallet is added to Daedalus
And both wallets are synced
And the Ledger Nano X wallet balance is >=2
When I access the "Send" screen for the Shelley wallet
And I paste an address from the Byron wallet on the "Receiver" input
And I insert a valid amount on the "Amount" input
And I click on Next"
And I plug, unlock and export the public key on the device
And confirm the transaction on the dialog
Then the transaction is performed
And the transaction can be visible on both wallets
```

**LEDGER NANO X - Scenario 3 - Redelegate Ledger Nano X Wallet**

```Gherkin
Given a Ledger Nano X wallet with at least 10 ADA is added
And the wallet is delegated
And both Daedalus and the wallet are synced
And the "Delegation Center" screen on "Staking Center" is displayed
When I click on "Redelegate" button for the wallet
And I follow the instructions on the wizard to redelegate the wallet
And I plug, unlock and export the public key on the device
And I confirm the delegation
Then the wallet is displayed as delegated to the new stake pool in the epoch number = current epoch + 2
```

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by ManusMcCole** [comment]
All scenarios passed on Trezor Model T 👍

**Review by ManusMcCole — APPROVED** [review]

**Review by gnpf — APPROVED** [review]

File: source/renderer/app/api/network/types.js [hunk 1]
```diff
@@ -1,7 +1,8 @@
 // @flow
 export type TipInfo = {
-  epoch: ?number,
-  slot: ?number,
+  epoch: number,
+  slot: number,
+  absoluteSlotNumber?: number,
```
Comment (nikolaglumac) [inline source/renderer/app/api/network/types.js:5]: @tomothespian please make this required attribute and extract it from Api response for both Node and Network tip 🙏

File: source/renderer/app/api/api.js [hunk 2]
```diff
@@ -2013,8 +2013,9 @@ export default class AdaApi {
         },
         networkTip: networkTip
           ? {
-              epoch: get(networkTip, 'epoch_number', null),
-              slot: get(networkTip, 'slot_number', null),
+              epoch: get(networkTip, 'epoch_number', 0),
+              slot: get(networkTip, 'slot_number', 0),
+              absoluteSlotNumber: get(networkTip, 'absolute_slot_number', 0),
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:2018]: @tomothespian please also extract this value from NodeTip too 🙏

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ Changelog
 
 ### Chores
 
+- Implemented dynamic TTL for Hardware Wallets transactions ([PR 2331](https://github.com/input-output-hk/daedalus/pull/2331))
```
Comment (nikolaglumac) [inline CHANGELOG.md:12]: @tomothespian please change to:

```
Implemented dynamic TTL calculation for hardware wallets transactions
```

File: source/renderer/app/api/utils/patchAdaApi.js [hunk 4]
```diff
@@ -53,6 +53,7 @@ export default (api: AdaApi) => {
         networkTip: {
           epoch: get(network_tip, 'epoch_number', null),
           slot: get(network_tip, 'slot_number', null),
+          absoluteSlotNumber: get(network_tip, 'absolute_slot_number', 0),
```
Comment (nikolaglumac) [inline source/renderer/app/api/utils/patchAdaApi.js:56]: @tomothespian please also extract this value from NodeTip too 🙏

File: source/renderer/app/config/txnsConfig.js [hunk 5]
```diff
@@ -1 +1,2 @@
 export const PENDING_TIME_LIMIT = 10 * 60 * 1000; // 10 minutes | unit: milliseconds
+export const TX_DURATION = 2 * 60 * 60; // 2 hours | unit: seconds (7200)
```
Comment (nikolaglumac) [inline source/renderer/app/config/txnsConfig.js:2]: @tomothespian please rename this variable to: `TIME_TO_LIVE` and change the value of it to: 48 hours.

File: source/renderer/app/stores/HardwareWalletsStore.js [hunk 6]
```diff
@@ -14,6 +14,7 @@ import {
   isTrezorEnabled,
   isLedgerEnabled,
 } from '../config/hardwareWalletsConfig';
+import { TX_DURATION } from '../config/txnsConfig';
```
Comment (nikolaglumac) [inline source/renderer/app/stores/HardwareWalletsStore.js:17]: @tomothespian please rename to `TIME_TO_LIVE `

File: source/renderer/app/stores/HardwareWalletsStore.js [hunk 7]
```diff
@@ -1738,6 +1739,13 @@ export default class HardwareWalletsStore extends Store {
     return type;
   };
 
+  _getTtl = () => {
```
Comment (nikolaglumac) [inline source/renderer/app/stores/HardwareWalletsStore.js:1742]: @tomothespian please add type declaration

File: source/renderer/app/stores/HardwareWalletsStore.js [hunk 8]
```diff
@@ -1738,6 +1739,13 @@ export default class HardwareWalletsStore extends Store {
...
+    const { networkTip } = this.stores.networkStatus;
+    const absoluteSlotNumber = get(networkTip, 'absoluteSlotNumber', 0);
```
Comment (nikolaglumac) [inline source/renderer/app/stores/HardwareWalletsStore.js:1744]: @tomothespian we need to handle the missing `absoluteSlotNumber` better...
Perhaps throw an error here?

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@tomothespian please implement the requested changes 🙏

File: source/renderer/app/config/txnsConfig.js (hunk 5 above)
Comment (nikolaglumac) [reply source/renderer/app/config/txnsConfig.js:2]: @rvl confirmed this should be set at 2h like you did @tomothespian. Great 🎉

**Comment by tomothespian** [comment]
@nikolaglumac code improved

**Review by nikolaglumac — APPROVED** [review]
Great work @tomothespian ❤️

**Review by ManusMcCole — APPROVED** [review]

**Comment by mioriohk** [comment]
LGTM. Good work @tomothespian . Tested on Ledger Nano X on build 16250

**Review by mioriohk — APPROVED** [review]

**Pull request merged**